### PR TITLE
MS SQL db Metricset (main PR)

### PR DESCRIPTION
**Pull request by sayden**
Main PR suffix refers to the fact that this PR contains the first metricset plus the _sql engine_ to do the queries concurrently. It's expected that follow PR with new metricset just contains the queries to fetch the telemetry information.

Continuation of https://github.com/elastic/beats/pull/9202 which grouped 4 different metricsets.

This is the single `db` metricset to facilitate reviewing.

`db` metricset is a subset of some of the most relevant information you can find [here](https://docs.microsoft.com/en-us/sql/relational-databases/system-dynamic-management-views/database-related-dynamic-management-views-transact-sql?view=sql-server-2017).

The way it has to work is: the metricset has the SQL queries it needs to get the information from the DB. Queries are sent to a central _engine_ created for each metricset in `x-pack/metricbeat/module/mssql` which run them concurrently, returning the results so that they can be reported. `io.Closer` is implemented on the _engine_ to close the db connection. I understood that it's called from the framework. Am I correct? @ruflin @jsoriano 

This concurrent structure was based on the fact that each fetch operation needs to establish a new connection to the database. Now that the connection is maintained active, a _transport_ channel could be used to report events and maybe the code would be a bit simpler (but I need to be sure about a couple of things of how the framework works before doing it to not leave dangling channels by accident)

## GA

* [X] Supported versions are documented
* [X] Supported operating systems are documented (if applicable)
* [X] Integration tests
* [X] System tests
* [X] Automated checks that all fields are documented
* [X] Documentation
* [X] Example `data.json` exists and an automated way to generate it exists (`go test -data`)
* [X] Fields follow naming conventions: https://www.elastic.co/guide/en/beats/devguide/master/event-conventions.html
* [ ] Dashboards exists (if applicable)
* [ ] Kibana Home Tutorial (if applicable)

### CODE REVIEW

**Comment by ruflin** [comment]
It's up to you to make sure the connections are properly closed. Best check the mysql metricset on how it's done there.

File: NOTICE.txt [hunk 1]
```diff
@@ -9,6 +9,15 @@ Third party libraries used by the Elastic Beats project:
 ==========================================================================
 
 
+--------------------------------------------------------------------
+Dependency: cloud.google.com/go
```
Comment (ruflin) [inline NOTICE.txt]: Could you check how much the binary size increases with these new dependencies?

File: x-pack/metricbeat/module/mssql/_meta/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,8 @@
+FROM microsoft/mssql-server-linux:2017-latest
```
Comment (ruflin) [inline x-pack/metricbeat/module/mssql/_meta/Dockerfile]: Nice that we have a docker image to test this too 👍

File: x-pack/metricbeat/module/mssql/_meta/fields.yml [hunk 3]
```diff
@@ -0,0 +1,11 @@
+- key: mssql
+  title: "mssql"
+  description: >
+    experimental[]
```
Comment (ruflin) [inline x-pack/metricbeat/module/mssql/_meta/fields.yml]: Should we make this beta directly?

File: x-pack/metricbeat/module/mssql/doc.go [hunk 4]
```diff
@@ -2,5 +2,5 @@
 // or more contributor license agreements. Licensed under the Elastic License;
 // you may not use this file except in compliance with the Elastic License.
 
-// Package foo is a Metricbeat module that contains MetricSets.
-package foo
+// Package mssql is a Metricbeat module that contains MetricSets.
```
Comment (ruflin) [inline x-pack/metricbeat/module/mssql/doc.go]: Was this moved on purpose? I would assume both files should exist?

File: x-pack/metricbeat/module/mssql/fetcher.go [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+	}
+}
+
+// doQueries is executed on object creation from the metricsets via NewFetcher.
+// It will launch a goroutine for each query stored and wait for all results
+// before returning.
+func (f *Fetcher) doQueries() {
+	f.wg.Add(len(f.queries))
+	for _, q := range f.queries {
+		go func() {
```
Comment (ruflin) [inline x-pack/metricbeat/module/mssql/fetcher.go]: I would suggest for now to not do this in parallel and run one query after the other until we hit a bottle neck. What is the maximum of queries we need to run here? I would expect only 1 or 2.

The reason I would remove it is to have a simpler code.

File: x-pack/metricbeat/module/mssql/_meta/fields.yml (hunk 3 above)
Comment (sayden) [reply x-pack/metricbeat/module/mssql/_meta/fields.yml]: I'd prefer to be conservative about this because there are lots of versions and operating systems involved, more than usual I mean because it's just not a Ubuntu or a Centos but also a MS Windows *

Comment (ruflin) [reply x-pack/metricbeat/module/mssql/_meta/fields.yml]: In general the meaning of Experimental at Elastic is we still might remove the feature in the future, beta means we still make breaking changes. I assume when we get it in I don't see a reason we would remove it again. To mitigate the problems around multiple platforms we should state very clearly what are the platforms we tested against.

File: x-pack/metricbeat/module/mssql/db/data.go [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+	schema = s.Schema{
+		"database": s.Object{
+			"id": c.Str("database_id"),
+		},
+		// Returns space usage information for the transaction log.
+		"log_space_usage": s.Object{
+			"total_bytes":             c.Int("total_log_size_in_bytes"),
+			"used_bytes":              c.Int("used_log_space_in_bytes"),
+			"used_percent":            c.Float("used_log_space_in_percent"),
+			"since_last_backup_bytes": c.Int("log_space_in_bytes_since_last_backup"),
```
Comment (jsoriano) [inline x-pack/metricbeat/module/mssql/db/data.go]: These fields could be
```
total.bytes
used.bytes
used.pct
since_last_backup.bytes
```
As in other modules.

File: x-pack/metricbeat/module/mssql/fetcher.go (hunk 5 above)
Comment (sayden) [reply x-pack/metricbeat/module/mssql/fetcher.go]: I agree on having a simpler code. To answer your question, the max number of queries I have seen to run really depends on the deep and breath of the module but I'd say around 7-10

File: x-pack/metricbeat/module/mssql/doc.go (hunk 4 above)
Comment (sayden) [reply x-pack/metricbeat/module/mssql/doc.go]: Sorry, I think this have been my git tool that detected a rename instead of an single addition because I really didn't do this manually. I will fix it in the next commit

File: NOTICE.txt (hunk 1 above)
Comment (sayden) [reply NOTICE.txt]: Master -> 72740520 bytes ~70mb
PR -> 72716960 bytes ~70mb

Comment (ruflin) [reply NOTICE.txt]: It seems with the PR the build is smaller? Not sure how this is possible. Perhaps new stuff in master and not rebase yet?

Comment (sayden) [reply NOTICE.txt]: You're were right, a rebase was needed. But after rebasing the sizes are now equals which doesn't make sense either :confused: 
Master -> 72740520
PR -> 72740520

Comment (ruflin) [reply NOTICE.txt]: how do you build it? is the import correct? Are you looking at the x-pack one?

Comment (sayden) [reply NOTICE.txt]: Actually, I was doing it in x-pack and the sizes were also equal, so I switched to OSS because I thought that there was something I missed from the current build system in X-Pack but the sizes were also equal...  I have had problems in the past where we needed to delete the contents in `$GOPATH/pkg` to solve some library issues but even with this the binary sizes are equal

File: x-pack/metricbeat/metricbeat.reference.yml [hunk 7]
```diff
@@ -493,6 +493,14 @@ metricbeat.modules:
   # Password to use when connecting to MongoDB. Empty by default.
   #password: pass
 
+#-------------------------------- Mssql Module --------------------------------
+- module: mssql
+  metricsets:
+    - "db"
+  hosts: ["sqlserver://sa@localhost"]
```
Comment (sayden) [inline x-pack/metricbeat/metricbeat.reference.yml]: NOTE: `sa` is the common name for _system administrator_ in MS SQL

File: x-pack/metricbeat/module/mssql/_meta/Dockerfile (hunk 2 above)
Comment (sayden) [reply x-pack/metricbeat/module/mssql/_meta/Dockerfile]: I'd like to work in this module on Windows too, setting up tests and versions to make it really reliable but I'd prefer to do it in a follow up PR.

File: x-pack/metricbeat/module/mssql/db/data.go (hunk 6 above)
Comment (sayden) [reply x-pack/metricbeat/module/mssql/db/data.go]: I did not know it, done :)

File: x-pack/metricbeat/module/mssql/fetcher.go (hunk 5 above)
Comment (sayden) [reply x-pack/metricbeat/module/mssql/fetcher.go]: Concurrency removed

File: x-pack/metricbeat/module/mssql/_meta/fields.yml (hunk 3 above)
Comment (sayden) [reply x-pack/metricbeat/module/mssql/_meta/fields.yml]: Replaced experimental mentions with beta. Also the `logp` function

File: NOTICE.txt (hunk 1 above)
Comment (ruflin) [reply NOTICE.txt]: The problem is that `_ github.com/elastic/beats/x-pack/metricbeat/include` is missing in the main.go file so imports are never loaded. That also means probably your testing is not working? I get the two following values:

```
Before: 57200688
After: 58519728
```

File: x-pack/metricbeat/module/mssql/db/data.go (hunk 6 above)
Comment (ruflin) [reply x-pack/metricbeat/module/mssql/db/data.go]: Here you can find the conventions: https://www.elastic.co/guide/en/beats/devguide/current/event-conventions.html

File: x-pack/metricbeat/module/mssql/_meta/Dockerfile (hunk 2 above)
Comment (jsoriano) [reply x-pack/metricbeat/module/mssql/_meta/Dockerfile]: Yes, leave windows support for a future change, I don't think we have Windows CI machines with native docker.

File: x-pack/metricbeat/module/mssql/fetcher.go [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+
+	for i, res := range rawResult {
+		if res != nil {
+			mapOfResults[r.columnNames[i]] = *res
+		}
+	}
+
+	result, err := f.schema.Apply(mapOfResults)
+	if err != nil {
+		r.reporter.Error(err)
```
Comment (jsoriano) [inline x-pack/metricbeat/module/mssql/fetcher.go]: Maybe it worths to wrap this error with a message about this failing when applying the schema.

File: x-pack/metricbeat/module/mssql/db/data.go (hunk 6 above)
Comment (sayden) [reply x-pack/metricbeat/module/mssql/db/data.go]: Favorited :heart_eyes:

File: x-pack/metricbeat/module/mssql/db/db.go [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+	fetcher *mssql.Fetcher
+}
+
+// New create a new instance of the MetricSet
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Beta("The mssql db metricset is beta.")
+
+	logger := logp.NewLogger("mssql.db").With("host", base.HostData().SanitizedURI)
+
+	fetcher, err := mssql.NewFetcher(base.HostData().URI,
```
[the remark this answers is not in the document]
Comment (sayden) [reply x-pack/metricbeat/module/mssql/db/db.go]: Do you mean apart from this one? https://github.com/elastic/beats/pull/9414/files#diff-c4de2e39351be98f70b39ad4e9dfa233R58

[the remark this answers is not in the document]
Comment (jsoriano) [reply x-pack/metricbeat/module/mssql/db/db.go]: Is this being called at some moment? If you implement a `Close()` method in the metricset it will be called when the metricset is stopped, the close in the metricset could call `m.fetcher.Close()`.

File: x-pack/metricbeat/module/mssql/db/db.go [hunk 10]
```diff
@@ -51,3 +51,8 @@ func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
 func (m *MetricSet) Fetch(reporter mb.ReporterV2) {
 	m.fetcher.Report(reporter)
 }
+
+// Closes the connection to the server at the engine level
+func (m *MetricSet) Close() error {
```
Comment (jsoriano) [inline x-pack/metricbeat/module/mssql/db/db.go]: :+1:

File: NOTICE.txt (hunk 1 above)
Comment (sayden) [reply NOTICE.txt]: Good catch @ruflin I think it's not surprising that the tests have passed because there are not system tests in xp/metricbeat yet (which is the one that compiles the binary) and the integration tests (the ones run with `mage goItengTest`) compiles at the package level. When testing in local I was working normally, with iterations of failing tests, fixing, etc.

So, the package itself is 9260 bytes in a single file https://github.com/elastic/beats/pull/9414/files#diff-563909ea0b3926d894389fca8b784bdb plus another 11358 of the ASL2 file.

Go source code is just importing `fmt` and `time` so I'd say that the increase in size comes 99% from the mssql library itself.

File: x-pack/metricbeat/module/mssql/db/db.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (sayden) [reply x-pack/metricbeat/module/mssql/db/db.go]: Now I understand. Somehow I was thinking that `Close` was being called only at the module level. Thanks for the clarification

File: x-pack/metricbeat/module/mssql/fetcher.go [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+	columnNames, err := rows.Columns()
+	if err != nil {
+		return errors.Wrap(err, "error getting column names")
+	}
+
+	rowsResult := newRowsResult(reporter, columnNames, rows)
+
+	if err = rowsResult.handle(f.schema); err != nil {
+		return errors.Wrap(err, "could not convert rows result to beats mapstr")
+	}
```
Comment (jsoriano) [inline x-pack/metricbeat/module/mssql/fetcher.go]: When I said to pass the row as argument and to keep all the iteration in the same place, I meant to do something like:
```
row := make(...)
for rows.Next() {
    if err = rows.Scan(row...); err != nil {
        ...
    }
    f.handleRow(reporter, columnNames, row)
}
```

File: x-pack/metricbeat/module/mssql/db/_meta/fields.yml [hunk 12]
```diff
@@ -0,0 +1,28 @@
+- name: db
+  type: group
+  description: >
+    db
+  fields:
+    - name: database
+      type: group
+      description: Database metricset
```
[the remark this answers is not in the document]
Comment (jsoriano) [reply x-pack/metricbeat/module/mssql/db/_meta/fields.yml]: My comment here was only about their descriptions, I don't understand why `db.database` description is "Database metricset" and the one for `db` is "db". I'd say that  `db` description should say something about the metricset metrics, and `db.database` something about database info, or just be empty.

Regarding organization, if `database` is going to be shared between metricsets as I think I suggested in some of the other PRs then yes, this should be placed at the module level, so the id is in `mssql.database.id`.

Notice that after applying an schema you can still modify the events.

**Comment by sayden** [comment]
jenkins, test this

File: x-pack/metricbeat/module/mssql/fetcher.go (hunk 11 above)
Comment (sayden) [reply x-pack/metricbeat/module/mssql/fetcher.go]: I'd like more the current change. You were right, there was a mix of responsibilities that makes the code a bit confusing. `Fetcher` which is a struct that _must fetch data from the database` was iteraring over the row results to get individual rows. Now it just calls `rowsResult.handle` so that the responsibility of extracting data from rows is fully abstracted from `Fetcher` into the `rowsResultHandler` which, in fact, could be in its own file even.

**Comment by sayden** [comment]
Rebased over https://github.com/elastic/beats/pull/9509

File: x-pack/metricbeat/module/mssql/fetcher.go (hunk 11 above)
Comment (jsoriano) [reply x-pack/metricbeat/module/mssql/fetcher.go]: I was not refering to separating responsibilities, but good point, actually this can be important in case we reuse this for other sql services.

The only thing now is that I see this handler struct a bit artificial:
* I'd expect for a handler to have a method like `handle(the thing to be handled)`
* I wouldn't expect the handler to keep state, why is `dest` a member of the struct and not just a local variable in `handle`? (well, this can be seen as some kind of optimization as the object also knows about the columns)
* I don't see a clear separation between the members of the handler struct and the inputs of the handle method. 
for example if handling here means _to apply an schema to generate an event and report it_ why does it contain the reporter, but not the schema?

I won't block this PR because of these questions, it can be a matter of code styles. And in any case this has to be revisited when we move this fetcher to be a reusable helper.

**Comment by sayden** [comment]
jenkins, test this

File: x-pack/metricbeat/module/mssql/db/data.go (hunk 6 above)
Comment (sayden) [reply x-pack/metricbeat/module/mssql/db/data.go]: I have just read [here](https://www.elastic.co/guide/en/beats/devguide/current/event-conventions.html) that we shouldn't use `.` in individual field names. Is it correct? What am I misunderstanding?

Comment (ruflin) [reply x-pack/metricbeat/module/mssql/db/data.go]: This is correct. But it means in the above you need to use something like `s.Object{ ...` like you can see here in the dev guide: https://www.elastic.co/guide/en/beats/devguide/current/creating-metricsets.html Also check other metricsets for examples.

File: x-pack/metricbeat/module/mssql/_meta/docs.asciidoc [hunk 13]
```diff
@@ -0,0 +1,2 @@
+This is the mssql module.
```
Comment (ruflin) [inline x-pack/metricbeat/module/mssql/_meta/docs.asciidoc]: I think the module description could be a bit more extensive.

File: x-pack/metricbeat/module/mssql/db/_meta/docs.asciidoc [hunk 14]
```diff
@@ -0,0 +1 @@
+This is the db metricset of the module mssql.
```
Comment (ruflin) [inline x-pack/metricbeat/module/mssql/db/_meta/docs.asciidoc]: Should we add some more details here on what metrics are fetched?

File: x-pack/metricbeat/module/mssql/db/_meta/fields.yml (hunk 12 above)
[the remark this answers is not in the document]
Comment (ruflin) [reply x-pack/metricbeat/module/mssql/db/_meta/fields.yml]: Agree with @jsoriano that as soon as we see more metricsets, we potentially should redicuss how we organise the fields and which ones are on the module level.

File: x-pack/metricbeat/module/mssql/fetcher.go [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+		queries: qs,
+		schema:  schema,
+		db:      db,
+		log:     log,
+	}
+
+	return f, nil
+}
+
+// Fetcher will make queries concurrently to the database to fetch results. It
```
Comment (ruflin) [inline x-pack/metricbeat/module/mssql/fetcher.go]: This needs update as it's not concurrently anymore.

File: x-pack/metricbeat/module/mssql/fetcher.go [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+	return f.db.Close()
+}
+
+// Report receives a mb.ReporterV2 to send the data to the outputs. Because the operations are common between metricsets
+// it can be extracted to a common function
+func (f *Fetcher) Report(reporter mb.ReporterV2) {
+	var err error
+	for _, q := range f.queries {
+		if err = f.getEventsWithQuery(f.db, q, reporter); err != nil {
+			logp.Error(err)
```
Comment (ruflin) [inline x-pack/metricbeat/module/mssql/fetcher.go]: Why are you passing the reporter to GetEventsWithQuery and further down instead of just reporting and error here on the top level the same as you do with the logging?

File: x-pack/metricbeat/module/mssql/fetcher.go [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+// to an slice of common.MapStr.
+func (f *Fetcher) getEventsWithQuery(db *sql.DB, query string, reporter mb.ReporterV2) error {
+	// Returns the global status, also for versions previous 5.0.2
+	rows, err := db.Query(query)
+	if err != nil {
+		return errors.Wrapf(err, "error performing db query='%v'", query)
+	}
+	defer func() {
+		if err = rows.Close(); err != nil {
+			logp.Error(errors.Wrap(err, "error closing query row"))
```
Comment (ruflin) [inline x-pack/metricbeat/module/mssql/fetcher.go]: Could we use a name return value here instead so we don't have to log it and still get the error reported as return value?

File: x-pack/metricbeat/module/mssql/fetcher.go (hunk 16 above)
Comment (jsoriano) [reply x-pack/metricbeat/module/mssql/fetcher.go]: I suggested that on the concurrent implementation to report the events as soon as they are generated, instead of accumulating them in a shared list that required mutexes.
I still think it can make sense to report the events as soon as they are generated, but it could be more discussed.

Comment (jsoriano) [reply x-pack/metricbeat/module/mssql/fetcher.go]: For context, the original thread https://github.com/elastic/beats/pull/9202/files#r235994447

Comment (jsoriano) [reply x-pack/metricbeat/module/mssql/fetcher.go]: One per row in the response to the query, for the case of the db metricset I guess this will be one per database, this would be something to be added to the documentation of the metricset.

Comment (ruflin) [reply x-pack/metricbeat/module/mssql/fetcher.go]: 1 per db make sense.

File: x-pack/metricbeat/tests/system/config/libbeat.yml.j2 [hunk 18]
```diff
@@ -0,0 +1,96 @@
+#================================ General =====================================
```
Comment (sayden) [inline x-pack/metricbeat/tests/system/config/libbeat.yml.j2]: This file is copy-pasted from https://github.com/elastic/beats/blob/master/metricbeat/tests/system/config/metricbeat.yml.j2

File: x-pack/metricbeat/tests/system/config/metricbeat.yml.j2 [hunk 19]
```diff
@@ -0,0 +1,124 @@
+###################### Metricbeat Configuration Template #######################
```
Comment (sayden) [inline x-pack/metricbeat/tests/system/config/metricbeat.yml.j2]: This file is copied from https://github.com/elastic/beats/blob/master/metricbeat/tests/system/config/metricbeat.yml.j2

File: x-pack/metricbeat/tests/system/config/metricbeat.yml.j2 [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+  - type: {{provider}}
+    {%- if settings %}
+    {%- for k, v in settings.items() %}
+    {{k}}: {{v | default([])}}
+    {%- endfor %}
+    {%- endif %}
+  {%- endfor %}
+{% endif %}
+
+{% include './tests/system/config/libbeat.yml.j2' %}
```
Comment (sayden) [inline x-pack/metricbeat/tests/system/config/metricbeat.yml.j2]: I couldn't make this import work without pasting the `libbeat.yml.j2` file directly into `x-pack/metricbeat/tests/system/config`.

File: x-pack/metricbeat/module/mssql/_meta/fields.yml [hunk 21]
```diff
@@ -0,0 +1,11 @@
+- key: mssql
+  title: "mssql"
+  description: >
+    beta[]
```
Comment (jsoriano) [inline x-pack/metricbeat/module/mssql/_meta/fields.yml]: Use `release: beta` instead of using a static description, and add this entry also to metricset fields file.

File: x-pack/metricbeat/tests/system/config/libbeat.yml.j2 (hunk 18 above)
Comment (jsoriano) [reply x-pack/metricbeat/tests/system/config/libbeat.yml.j2]: We should parametrize the [template loader](https://github.com/elastic/beats/blob/master/libbeat/tests/system/beat/beat.py#L301) in a different way so we can reuse these files.
I am ok with duplicating these files by now to not block this.

File: x-pack/metricbeat/module/mssql/fetcher.go (hunk 11 above)
Comment (sayden) [reply x-pack/metricbeat/module/mssql/fetcher.go]: I'll write down this comments to have a look in a following PR. I understand your point. Thanks Jaime :)

File: x-pack/metricbeat/module/mssql/_meta/Dockerfile (hunk 2 above)
Comment (ruflin) [inline x-pack/metricbeat/module/mssql/_meta/Dockerfile]: One additional thing I just spotted here: We normally defined the depenendcy up to the bug fix release to make sure CI always uses the exact same version as we had issues in the past. Not sure if we have here a more fine grained version or if this is already it.

File: x-pack/metricbeat/module/mssql/db/data.go [hunk 22]
```diff
@@ -0,0 +1,25 @@
...
+)
+
+var (
+	schema = s.Schema{
+		"database": s.Object{
+			"id": c.Str("database_id"),
+		},
+		// Returns space usage information for the transaction log.
+		"log_space_usage": s.Object{
+			"total.bytes":             c.Int("total_log_size_in_bytes"),
```
Comment (ruflin) [inline x-pack/metricbeat/module/mssql/db/data.go]: You need to use the `s.Object` here to have in the end `total` as an object and `bytes` as the key. Same as you do on line 18.

File: x-pack/metricbeat/module/mssql/fetcher.go (hunk 17 above)
Comment (sayden) [reply x-pack/metricbeat/module/mssql/fetcher.go]: Oh! Good catch! Because I was actually ofuscating `err` from any of the other calls in case of an error here. So, as far as I can see it could be done like this (but it's ugly as hell):

```
func (f *Fetcher) getEventsWithQuery(db *sql.DB, query string, reporter mb.ReporterV2) (err error) {
	var rows *sql.Rows
	rows, err = db.Query(query)
	if err != nil {
		return errors.Wrapf(err, "error performing db query='%v'", query)
	}
	defer func() {
		if err2 := rows.Close(); err != nil {
			if err != nil {
				err = errors.Wrap(err, err2.Error())
			} else {
				err = err2
			}
		}
	}()
```

File: x-pack/metricbeat/module/mssql/_meta/Dockerfile (hunk 2 above)
Comment (sayden) [reply x-pack/metricbeat/module/mssql/_meta/Dockerfile]: Actually, there's a `GA` version. I have already changed it

File: x-pack/metricbeat/module/mssql/fetcher.go [hunk 23]
```diff
@@ -0,0 +1,155 @@
...
+	}
+	defer func() {
+		if err2 := rows.Close(); err != nil {
+			if err != nil {
+				err = errors.Wrap(err, err2.Error())
+			} else {
+				err = err2
+			}
+		}
+	}()
```
Comment (jsoriano) [inline x-pack/metricbeat/module/mssql/fetcher.go]: Sorry, but as you said this is actually a bit ugly :grimacing: Also [according to documentation](https://golang.org/pkg/database/sql/#Rows.Close) rows are closed automatically by Next, so this should only be needed if the Scan/Next loop is unexpectedly interrupted (having rows creation and the Scan/Next loop in the same place would be another reason in favour of [this proposal](https://github.com/elastic/beats/pull/9414#discussion_r241210948)).

Also for better or worse it is quite usual to ignore the error on calls to `Close()`, even in the examples in the documentation for the `sql` module... I think we'd be fine with reducing this to `defer Close()`, or at much log this error at the debug level, because an error here is probably very very rare.

I'd change it here, but we will have to revisit this code in any case, so as you prefer... but if you keep this here, please at least use a more meaningful name for the inner error, like `closeErr`.

Comment (sayden) [reply x-pack/metricbeat/module/mssql/fetcher.go]: I think we should not block this PR with more stylish and subjective things.

A very very rare errors that is hidden on purpose leads to a very very difficult to debug, discover and fix bugs so also very very difficult to maintain. That's way I prefer explicit approaches (say ugly if you want) Than just to obfuscate the error.

So please, if you don't mind, let's move this PR forward because it has been hangling for months already, it has 81 comments and there's still has a lot of work to do with the upcoming metricsets that will need a lot of review in many aspects.

File: x-pack/metricbeat/module/mssql/db/_meta/fields.yml (hunk 12 above)
[the remark this answers is not in the document]
Comment (jsoriano) [reply x-pack/metricbeat/module/mssql/db/_meta/fields.yml]: @sayden I think these descriptions should be fixed on this PR

File: x-pack/metricbeat/module/mssql/fetcher.go (hunk 23 above)
Comment (ruflin) [reply x-pack/metricbeat/module/mssql/fetcher.go]: I think the problem is that code that gets merge gets rarely changed again. My main issue with the above is that when I read the code, I don't really understand what it's doing above.

To keep this moving the following suggestion: Please add it to the module meta issue as an entry that has to be looked at again so we don't forget about the cleanup.

File: x-pack/metricbeat/module/mssql/_meta/docs.asciidoc [hunk 24]
```diff
@@ -0,0 +1,19 @@
+This is the https://www.microsoft.com/en-us/sql-server/sql-server-2017[Microsoft SQL 2017] Metricbeat module. It is still in beta and under active development to add new Metricsets and introduce enhancements.
+
+The idea is to improve the current set of metrics by fetching data from https://docs.microsoft.com/en-us/sql/relational-databases/system-dynamic-management-views/system-dynamic-management-views?view=sql-server-2017[System Dynamic Management Views]. A curated dashboard will also be included soon.
```
Comment (ruflin) [inline x-pack/metricbeat/module/mssql/_meta/docs.asciidoc]: I think this info belongs into the Meta Github issue and not the docs. The docs should only contain what we have not what we plan.

File: x-pack/metricbeat/module/mssql/fetcher.go (hunk 23 above)
Comment (sayden) [reply x-pack/metricbeat/module/mssql/fetcher.go]: I'm happy with any proposal that doesn't involve a function with 4 or 5 responsibilities :1st_place_medal: In any case, I already have some notes for the next PR that I'm planning to write in the header of the upcoming metricset PR once it's open. I already have to change few stylish things also in this particular file so we have time to discuss a way that is _easy for everybody_ to understand / look :smile:

File: x-pack/metricbeat/module/mssql/fetcher.go (hunk 8 above)
Comment (sayden) [reply x-pack/metricbeat/module/mssql/fetcher.go]: Done :)

File: x-pack/metricbeat/module/mssql/_meta/docs.asciidoc (hunk 13 above)
Comment (sayden) [reply x-pack/metricbeat/module/mssql/_meta/docs.asciidoc]: Done :)

File: x-pack/metricbeat/module/mssql/db/_meta/docs.asciidoc (hunk 14 above)
Comment (sayden) [reply x-pack/metricbeat/module/mssql/db/_meta/docs.asciidoc]: I can but it's going to be redundant with the in `fields.yml`, right? I want to be sure that you mean that information

File: x-pack/metricbeat/module/mssql/fetcher.go (hunk 15 above)
Comment (sayden) [reply x-pack/metricbeat/module/mssql/fetcher.go]: Solved

File: x-pack/metricbeat/module/mssql/_meta/fields.yml (hunk 21 above)
Comment (sayden) [reply x-pack/metricbeat/module/mssql/_meta/fields.yml]: Done

File: x-pack/metricbeat/module/mssql/db/data.go (hunk 22 above)
Comment (sayden) [reply x-pack/metricbeat/module/mssql/db/data.go]: Also done, but I'd like some review

File: x-pack/metricbeat/module/mssql/db/_meta/docs.asciidoc (hunk 14 above)
Comment (ruflin) [reply x-pack/metricbeat/module/mssql/db/_meta/docs.asciidoc]: Yes. I should rephrase this to not necessarly about the metrics but that it describes it fetches metrics about DB and also potentially links to MSSQL docs where these metrics are. We can do this in a follow up.

File: x-pack/metricbeat/module/mssql/db/data.go (hunk 22 above)
Comment (ruflin) [reply x-pack/metricbeat/module/mssql/db/data.go]: LGTM

**Comment by sayden** [comment]
The follow up actions of this PR are in this meta-issue about the entire module https://github.com/elastic/beats/issues/8698 including the docs reviewing, the sql row close action error handling

**Comment by sayden** [comment]
jenkins, test this please

**Comment by sayden** [comment]
Trying to figure out what's the Travis error of `cat: build/TEST-go-integration.out: No such file or directory`

**Comment by sayden** [comment]
jenkins, test this please

**Comment by ruflin** [comment]
@sayden Failures on CI are not related.
